Add skin mounting flow

smallketchup82 opened the pull request:
https://github.com/user-attachments/assets/e32c8b5c-92a3-45da-b1f5-837f07cac6a8

Similarly to the analogous option in beatmap editor, this allows arbitrary external modifications to skins. In particular this means that it is now possible to change gameplay assets without exporting and re-emporting the skin.

---

This pull request expands on the [mounting logic](https://github.com/ppy/osu/pull/28800) from beatmaps and integrates it with the skin editor, thereby introducing external editing functionality for skins.

I have applied my current knowledge of realm and threading to ensure the code is at the very least presentable. However, I acknowledge that further refinement and evaluation may be necessary. I am open to any feedback and suggestions to improve this implementation.

This is intended to be a stopgap solution until [Skin Editor gains support for editing skin files visually](https://github.com/ppy/osu/issues/18902#issuecomment-2137613764).

smallketchup82 committed:
Initial implementation

smallketchup82 committed:
Fix changes not being applied instantly, Improve import performance, lay down framework for abstracting logic

smallketchup82 committed:
Abstract out logic to SkinImporter

smallketchup82 committed:
Add a toggle for checking overwriting

The GetFile method in AddFile has a huge overhead, given we're doing
this in a loop.

Since we clear the files in the skin, we already know there won't be any
existing files, so we can skip all of that logic

smallketchup82 committed:
Make everything translatable

smallketchup82 committed:
Rename oldskin

smallketchup82 commented on osu.Game/Database/ModelManager.cs:
```diff
@@ -81,16 +81,19 @@ public void ReplaceFile(RealmNamedFileUsage file, Stream contents, Realm realm)
-        /// Add a file from within an ongoing realm transaction. If the file already exists, it is overwritten.
+        /// Add a file from within an ongoing realm transaction. If the file already exists, it is overwritten so long as <paramref name="overwrite"/> is true.
         /// </summary>
-        public void AddFile(TModel item, Stream contents, string filename, Realm realm)
```
As for why I touched this

*From [`b7883f1` (#30226)](https://github.com/ppy/osu/pull/30226/commits/b7883f18be12d304fddb57269c5bcb123e0763b2)*

> The GetFile method in AddFile has a huge overhead, given we're doing
> this in a loop.
> 
> Since we clear the files in the skin, we already know there won't be any
> existing files, so we can skip all of that logic

peppy replied:
This makes no sense and offers zero safeties.

If the thing has too high overhead, we fix later. Not as part of a huge change. Remove this commit please.

smallketchup82 commented on osu.Game/Database/RealmArchiveModelImporter.cs:
```diff
@@ -188,7 +188,9 @@ public async Task<ExternalEditOperation<TModel>> BeginExternalEditing(TModel mod
-            foreach (var realmFile in model.Files)
+            // Detach files from the model to avoid realm contention when copying to the external location.
+            // This is safe as we are not modifying the model in any way.
+            foreach (var realmFile in model.Files.Detach())
```
As  the comment states, I ran into an issue where detaching the skin doesn't actually end up detaching the files, resulting in a `Realms.Exceptions.RealmException: Realm accessed from incorrect thread.` exception.

I couldn't find a way to fix it apart from this. Fortunately, we only access the realm files here to mount them, so we aren't modifying the files in a way that requires them to be attached.

bdach replied:
Two questions:

- Where / when is the "skin" being "detached"?
- Does detach even... _work_ for `SkinInfo`? It sure isn't _registered_ in the automapper detach machinery.

https://github.com/ppy/osu/blob/74675c8c7858416a241e29804ba88ca1c1f5e476/osu.Game/Database/RealmObjectExtensions.cs#L171-L179

Maybe adding `SkinInfo` in there just fixes this. I dunno.

smallketchup82 replied:
The skin is being detached [here](https://github.com/smallketchup82/osu/blob/15c7a12174687d5bcb9d058bcbb72a35e3a42b35/osu.Game/Overlays/SkinEditor/SkinEditor.cs#L283)

`CurrentSkin.Value.SkinInfo` in `SkinEditor` returns a `Live<SkinInfo>`. My original idea for fixing the issue with "realm being accessed from an incorrect thread" was to detach the `Live<SkinInfo>` so that we're only working with a `SkinInfo`. Though, simply stopping here (without detaching `Files`) doesn't work and results in the thread error. Adding in the `.Detach()` fixes it.

I got here via trial and error so I don't know for certain. But my best guess is that it likely has to do with `RealmNamedFileUsage` being in the automapper detach machinery in which you specified. Maybe it's only detaching the `Files` that ends up fixing the issue, not necessarily the detaching of the `Live<SkinInfo>`.

At any rate, I guess detaching the `Live<SkinInfo>` serves only to convert from `Live<SkinInfo>` to `SkinInfo`. But a quick test without detaching the `SkinInfo` seems to work fine. So maybe that's unnecessary?

bdach replied:
self-resolved via https://github.com/ppy/osu/pull/30226/commits/2e0e7ff3c2c21c8718fae9ffcf954f5fb9b10e96

smallketchup82 commented on osu.Game/Overlays/SkinEditor/SkinEditor.cs:
```diff
@@ -274,6 +282,61 @@ protected override void LoadComplete()
+
+                // Create a new skin instance to ensure the skin is reloaded
+                // If there's a better way to reload the skin, this should be replaced with it.
+                currentSkin.Value = newSkinInfo.CreateInstance(skins);
```
This is where most of my confidence in the code quality of this PR declines. This was the only way I found that would refresh the skin instantly. I looked through at least 15 different classes, and even tests, to figure out a proper way of reloading skins. Was left empty handed and came up with this which works. The reason I feel unconfident about it is because it feels rather hacky.

smallketchup82 commented on osu.Game/Skinning/SkinImporter.cs:
```diff
@@ -43,6 +46,35 @@ public SkinImporter(Storage storage, RealmAccess realm, IStorageResourceProvider
+            skinInfoLive.PerformWrite(skinInfo =>
+            {
+                // Not sure if this deletes the files from the storage or just the database.
+                skinInfo.Files.Clear();
```
As the comment states. Does this also flag the files for deletion? I tried to test it but couldn't conclude whether it does or doesn't. And `ModelManager` seems to do roughly the same in its [`DeleteFile`](https://github.com/ppy/osu/blob/b7883f18be12d304fddb57269c5bcb123e0763b2/osu.Game/Database/ModelManager.cs#L70) method. So I'm a bit confused here.

peppy replied:
Don't add comments for this. Ask elsewhere.

And yes, it does. Storage is managed as a separate layer based on file references.

smallketchup82 reviewed (commented):
Going to leave some comments here for reviewers on why I did certain things and any concerns I have. Apologies if I shouldn't be doing it this way, please correct me for future reference if so.

smallketchup82 commented:
Forgot tests, sorry about that. Will add them in ASAP

peppy commented:
> Forgot tests, sorry about that. Will add them in ASAP

Seems back to front. Like you should be creating the tests to develop the feature. Adding as an afterthought is mostly pointless for something like this? But let's see what you come up with.

smallketchup82 committed:
Revert "Add a toggle for checking overwriting"

This reverts commit b7883f18be12d304fddb57269c5bcb123e0763b2.

peppy commented:
I don't like this "simple" approach. Maybe simple for you, but not for a user. And also means you could potentially exit the skin editor or make edits in it while mounted, then overwrite those edits.

There should be a big button for the user to click once done external editing.

smallketchup82 commented:
I can look into making an overlay equivalent of `ExternalEditScreen`, though this might take me 1 or more weeks given my current schedule.

I'm not very experienced in overlay & screen logic yet, so I'm not sure, but would it be possible to reuse `ExternalEditScreen` for this?

smallketchup82 committed:
Merge branch 'master' into skin-mounting

smallketchup82 commented:
Sorry, misclick

smallketchup82 committed:
Remove a comment

smallketchup82 committed:
Switch to using an overlay interface for skin mounting

smallketchup82 commented:
I've managed to get an overlay component working. It's more or less a port of `ExternalEditScreen` to an overlay. With a bit of work it can be made compatible with beatmaps, and `ExternalEditScreen` can be replaced with it, so that the beatmap editor uses the overlay as well.

Here's a video demonstrating what the flow looks like now:

https://github.com/user-attachments/assets/e32c8b5c-92a3-45da-b1f5-837f07cac6a8

smallketchup82 committed:
Switch overlay colour scheme to purple

Figured this made sense since we're using purple buttons. It doesn't
really seem to change anything visually though

peppy commented on osu.Game/Overlays/ExternalEditOverlay.cs:
```diff
@@ -37,7 +37,7 @@ public partial class ExternalEditOverlay : OsuFocusedOverlayContainer
 
         [Cached]
-        private readonly OverlayColourProvider colourProvider = new OverlayColourProvider(OverlayColourScheme.Pink);
+        private readonly OverlayColourProvider colourProvider = new OverlayColourProvider(OverlayColourScheme.Purple);
```
It should inherit from the skin editor's colours, no?

smallketchup82 committed:
Switch colour scheme to blue

peppy commented on osu.Game/Skinning/SkinImporter.cs:
```diff
@@ -43,6 +46,34 @@ public SkinImporter(Storage storage, RealmAccess realm, IStorageResourceProvider
+        {
+            var skinInfoLive = original.ToLive(Realm);
+
+            skinInfoLive.PerformWrite(skinInfo =>
+            {
+                skinInfo.Files.Clear();
+
+                string[] filesInMountedDirectory = Directory.EnumerateFiles(task.Path, "*.*", SearchOption.AllDirectories).Select(f => Path.GetRelativePath(task.Path, f)).ToArray();
+
+                foreach (string file in filesInMountedDirectory)
+                {
+                    using var stream = File.OpenRead(Path.Combine(task.Path, file));
+
+                    modelManager.AddFile(original, stream, file);
+                }
+            });
```
The process should probably match the beatmap reimport one, optimally.

Maybe it can shortcut in the case that ini/json files didn't change, if the full import takes too long (although I'd like to hope it doesn't).

smallketchup82 replied:
I was hoping that it would update automatically, since changing the skin.ini and creating a new skin instance for the `SkinInfo` sounds like it should update the skin's name in realm, but it probably isn't because we're reusing the old `SkinInfo` which would have the old `Name` property.

I'll probably end up implementing it as peppy suggested. Before processing the other files, I'll hash the two skin.ini's. If there's a change, I'll investigate what changed and reflect them in realm. If nothing was changed, we're good and we can skip that entire process.

smallketchup82 replied:
I've finally gotten to fixing this, but it's now fixed. See [`165afe3` (#30226)](https://github.com/ppy/osu/pull/30226/commits/165afe357f5a52050f2337e54478f87739f125fb) for the details.

smallketchup82 committed:
Rename SkinInfo when it is changed in skin.ini

Peppy spoke about using a shortcut and/or hashes to determine if the
skin.ini is changed, and if so, then to rename the skin.

In my opinion, hashing and doing numerous comparisons is probably less
efficient than just syncing the SkinInfo's name during the update.

This is an easy solution that does what it needs to.

smallketchup82 commented:
Since the last review has been addressed, this should be good to go for another round.

bdach committed:
Merge branch 'master' into skin-mounting

bdach commented:
@smallketchup82 resolved conflicts for you, please double check when able

bdach reviewed (changes requested):
The settings menu can be opened when the skin is externally mounted. Meaning you can change skin while the skin is externally mounted. Doing this results in fundamentally broken behaviour and it should not even be permissible to enter such a state.

smallketchup82 committed:
Disallow opening settings menu when external edit ovelay is open

Also disallows using the random skin keybind when the external edit
overlay is open. SkinEditor should already be disabling it, but I
figured we might as well add this in for redundancy

smallketchup82 commented:
Thanks for catching that! I've fixed it in https://github.com/ppy/osu/pull/30226/commits/7a5e613cf68484876bbf0c4f6580b29127adb83b. Not sure if that's how or where you wanted it to be done, but it works well and disallows opening the settings menu when the external edit overlay is visible.

bdach commented on osu.Game/Overlays/ExternalEditOverlay.cs:
I heavily dislike basically the entirety of https://github.com/ppy/osu/pull/30226/commits/7a5e613cf68484876bbf0c4f6580b29127adb83b and I think the root cause of all that is this class. Why is this an overlay at all? Why is it not a blocking screen [like the editor implementation of this is](https://github.com/ppy/osu/blob/448790a76f9536b3a7e05480afc5b5e2b60bb548/osu.Game/Screens/Edit/ExternalEditScreen.cs#L31)?

Making the external skin edit thing a blocking screen should fix most of the problems that commit is trying to special-case. With the exception of the random skin hotkey, probably - that one I would argue needs better high-level handling, via the current skin bindable getting disabled or something. And even the commit message says the random thing handling may not be necessary at all...?

smallketchup82 replied:
I created this as an overlay because I initially designed it to work around the skin editor. To me, an overlay on top of another overlay seems simpler than hiding the skin editor overlay, pushing the external edit screen, and then showing the skin editor overlay once that’s done, or any other method that involves screen usage. This was a choice I made for simplicity. My entire idea was to add the overlay version of the external edit screen for the skin editor, and then move the beatmap editor to use the overlay version in a future PR.

To be blunt, if your only issue is https://github.com/ppy/osu/commit/7a5e613cf68484876bbf0c4f6580b29127adb83b, I can revert that. It was an addition to future-proof the external edit overlay (for example, if you mount a skin outside the context of the skin editor). However, as the commit description states, it’s optional and redundant since the skin editor already performs the same task.

peppy replied:
The idea of pushing a screen (if we're assuming the screen will be pushed to the game's main stack) likely isn't going to work well with gameplay, which is one of the common cases for the skin editor.

bdach replied:
Well then I don't know what the solution is, but I stand rather unmoved in opposition to `OsuGame` having to coordinate these two specific pieces deep inside the game. It feels not good. `OsuGame` is the last resort for stuff like that.

peppy replied:
I haven't looked into the PR code yet, but can agree that it definitely should not live in `OsuGame`.

smallketchup82 replied:
Sorry for the delay in response, I've been busy studying for exams.

> At least the weird explicit "suppress settings being opened if the external edit overlay is visible" logic would need to remain, would it not?

Yes, that's my bad. You're right here.

> Did you actually try it?

On a local branch yes. It suffers the issue that peppy mentioned (opening skin editor then pressing external edit mid gameplay takes you outside of gameplay). This happens even when paused which takes a good amount of usability away from this feature.

> but I stand rather unmoved in opposition to `OsuGame` having to coordinate these two specific pieces deep inside the game. It feels not good

I have no issue with moving it out of `OsuGame`, the question I have is where? Should I move this settings panel side? Or overlay side (i.e. DI settings panel to `ExternalEditOverlay` and disable it within `ExternalEditOverlay`)? A bit of direction here would be useful.

bdach committed:
Merge branch 'master' into skin-mounting

bdach committed:
Move external edit overlay to more proper namespace

bdach committed:
Eliminate weird parameter passing

bdach committed:
Use better method of disallowing skin changes during external edit

bdach committed:
Ensure pending changes are saved before initiating external edit operation

bdach committed:
Prevent hiding skin editor during external edit operation

bdach committed:
Fix error code paths just completely falling apart

You can't hide a drawable outside of update thread.

bdach committed:
Fix dodgy threading

The fact that the stuff "just worked" previously due to one load-bearing
detach in a random location is really scary because a lot of this was
just not written the way it is supposed to be.

peppy reviewed (commented):
Fine apart from proposed change.

bdach committed:
Move external skin edit overlay out of `OsuGame`

bdach commented on osu.Game/OsuGame.cs:
```diff
@@ -1227,7 +1225,7 @@ protected override void LoadComplete()
             loadComponentSingleFile(wikiOverlay = new WikiOverlay(), overlayContent.Add, true);
             loadComponentSingleFile(skinEditor = new SkinEditorOverlay(ScreenContainer), overlayContent.Add, true);
-            loadComponentSingleFile(externalEditOverlay = new ExternalEditOverlay(), overlayContent.Add, true);
+            loadComponentSingleFile(new ExternalEditOverlay(), overlayContent.Add, true);
```
applied with adjustments (disposal of the `Disposable` that `RegisterBlockingOverlay()` returns) in https://github.com/ppy/osu/pull/30226/commits/2c22158dbdaa13f3e27622e1df4317536e94300e
